src/main/java/htsjdk/variant/vcf/VCFCompoundHeaderLine.java

```diff
@@ -101,153 +257,79 @@ public int getCount(final VariantContext vc) {
-            count = 0;
-            if (GeneralUtils.DEBUG_MODE_ENABLED) {
-                System.err.println("FLAG fields must have a count value of 0, but saw " + count + " for header line " + getID() + ". Changing it to 0 inside the code");
+        // Note: this can drop extra attributes
+        if (!line1.equalsExcludingExtraAttributes(line2)) {
+            if (line1.getType().equals(line2.getType())) {
+                // The lines are different in some way, but have a common type.
+                // The Number entry is an Integer that describes the number of values that can be
+                // included with the INFO field. For example, if the INFO field contains a single
+                // number, then this value should be 1. However, if the INFO field describes a pair
+                // of numbers, then this value should be 2 and so on. If the number of possible
+                // values varies, is unknown, or is unbounded, then this value should be '.'.
+                conflictWarner.warn(line1, "Promoting header field Number to . due to number differences in header lines: " + line1 + " " + line2);
+                newLine = compoundHeaderLineResolver.apply(line1, line2);
+            } else if (line1.getType() == VCFHeaderLineType.Integer && line2.getType() == VCFHeaderLineType.Float) {
+                // promote key to Float
+                conflictWarner.warn(line1, "Promoting Integer to Float in header: " + line2);
+                newLine = line2;
+            } else if (line1.getType() == VCFHeaderLineType.Float && line2.getType() == VCFHeaderLineType.Integer) {
+                // promote key to Float
+                conflictWarner.warn(line1, "Promoting Integer to Float in header: " + line2);
+            } else {
+                throw new IllegalStateException("Incompatible header types, collision between these two types: " + line1 + " " + line2);
             }
         }
-    }
-
-    /**
-     * make a string representation of this header line
-     * @return a string representation
-     */
-    @Override
-    protected String toStringEncoding() {
-        Map<String, Object> map = new LinkedHashMap<String, Object>();
-        map.put("ID", name);
-        Object number;
-        switch (countType) {
-            case A:
-                number = VCFConstants.PER_ALTERNATE_COUNT;
-                break;
-            case R:
-                number = VCFConstants.PER_ALLELE_COUNT;
-                break;
-            case G:
-                number = VCFConstants.PER_GENOTYPE_COUNT;
-                break;
-            case UNBOUNDED:
-                number = VCFConstants.UNBOUNDED_ENCODING_v4;
-                break;
-            case INTEGER:
-            default:
-                number = count;
+        if (!line1.getDescription().equals(line2.getDescription())) {
+            conflictWarner.warn(line1, "Allowing unequal description fields through: keeping " + line2 + " excluding " + line1);
```

Comment:
I prefer not to keep duplicate IDs either, but it should be a way to choose which line description is preferred and/or merge both descriptions.

It is true that the sencond option can end in a concatenation of similar descriptions with period/no-period at the end of the line. But I believe that it is better than using one "randomly picked" description if they are not the same at all, leading to the same ID referring to different things depending on the variant/genotype/etc.

For example, An user looking at a concrete ID description as following  "x-statistic OR y-statistic" will catch that there is different information encoded in the same ID; otherwise, they will think that the value represents only the x-statistic.